R/calculate_tdm.R

```diff
@@ -0,0 +1,159 @@
+#'   scen_alloc_wt_tech_prod = 1:3,
+#'   plan_carsten = rep(0.5, 3)
+#' )
+#' pacta_results
+#'
+#' calculate_tdm(pacta_results, start_year = 2020)
+calculate_tdm <- function(data, start_year, additional_groups = NULL) {
+  check_calculate_tdm(data, start_year)
+
+  portfolio_weight <- filter(data, .data$allocation == "portfolio_weight")
+  if (nrow(portfolio_weight) == 0) {
+    return(warn_zero_rows(tdm_prototype()))
+  }
+
+  groups <- unique(c(crucial_tdm_groups(), additional_groups))
+
+  joint <- left_join(
+    filter(portfolio_weight, .data$year == start_year),
+    technology_level_dy(portfolio_weight, start_year, groups),
+    by = groups
+  )
+
+  # TODO: Try to extract one or more meaningful functions
+  joint %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    ungroup(.data$technology) %>%
+    add_tdm_sec(start_year) %>%
+    ungroup() %>%
+    select(names(tdm_prototype()), all_of(groups))
+}
+
+check_calculate_tdm <- function(data, start_year) {
+  stopifnot(is.data.frame(data), is.numeric(start_year))
+
+  crucial <- c(
+    "allocation",
+    crucial_tdm_groups(),
+    "year",
+    "scen_alloc_wt_tech_prod",
+    "plan_alloc_wt_tech_prod",
+    "plan_carsten"
+  )
+  check_crucial_names(data, crucial)
+
+  invisible(data)
+}
+warn_zero_rows <- function(data) {
+  # TODO: This function will only work if the allocation method is
+  # portfolio_weight ownership_weight outputs 0 for all carsten metric values,
+  # and thus we would be dividing by zero otherwise...
+  message <- 'Filtering for "portfolio_weight" allocation, outputs 0 rows'
+  warn(message, class = "has_zero_rows")
+
+  # Pass `data` to inline inside return()
+  invisible(data)
+}
+
+tdm_prototype <- function() {
+  tibble(
+    technology = character(0),
+    ald_sector = character(0),
+    tdm_tech = numeric(0),
+    tdm_sec = numeric(0),
+    reference_year = integer(0),
+  )
+}
+
+crucial_tdm_groups <- function() {
+  c("technology", "ald_sector")
+}
+
+#' Technology-level "dy"
+#' TODO: Explain the meaning of "dy".
+#' @noRd
+technology_level_dy <- function(data, start_year, groups) {
+  long <- data %>%
+    filter(.data$year %in% c(start_year, start_year + 5, start_year + 10)) %>%
+    add_time_step(start_year) %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    select(
+      !!!rlang::syms(groups),
+      .data$time_step,
+      scen_alloc = "scen_alloc_wt_tech_prod",
+      plan_alloc = "plan_alloc_wt_tech_prod"
+    )
+
+  long %>%
+    FIXME_pivot_wider(c("scen_alloc", "plan_alloc")) %>%
+    add_tdm_tech() %>%
+    select(.data$tdm_tech, all_of(groups)) %>%
+    distinct() %>%
+    ungroup()
+}
+
+FIXME_pivot_wider <- function(data, columns) {
+  # This avoids an error but may be the wrong thing to do
+  data %>%
+    pivot_wider(
+      names_from = .data$time_step,
+      values_from = all_of(columns),
+      values_fn = list # Suppress "Values are not uniquely identified"
+    ) %>%
+    unnest(cols = unlist(lapply(columns, contains)))
+}
```

Comment:
@jdhoffa please judge if this is a good way to handle the error you noted recenlty. We discussed the idea of catching the error and re-throwing it with a more meaningful message. But we didn't discuss the option I implemented -- by passing the error. Does this bypass leave us in a good place or should be embrace the error and stop execution?